### PR TITLE
Fix create table error

### LINKED ISSUE #4596: Fail to create table with some TiDB table info
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

A simplified case from https://asktug.com/t/topic/662946

```mysql
> set sql_mode='';
> CREATE TABLE test ( id int(16) NOT NULL AUTO_INCREMENT, PRIMARY KEY (id));
> alter table test add column b longblob not null default '';

-- p.s. Note that `set sql_mode=""` is required or the SQL will get blocked like
> CREATE TABLE test ( id int(16) NOT NULL AUTO_INCREMENT, PRIMARY KEY (id));
> alter table test add column b longblob not null default '';
ERROR 1101 (42000): BLOB/TEXT/JSON column 'b' can't have a default value
```

Those SQLs will create a table with JSON info as below. Note that the column type of "b" is not null long blob, but its "default"/"origin_default" field is `""` and "type.FLen" is `-1`.

```
{
	"cols": [{
		"comment": "",
		"default": null,
		"default_bit": null,
		"id": 1,
		"name": {
			"L": "id",
			"O": "id"
		},
		"offset": 0,
		"origin_default": null,
		"state": 5,
		"type": {
			"Charset": "binary",
			"Collate": "binary",
			"Decimal": 0,
			"Elems": null,
			"Flag": 515,
			"Flen": 16,
			"Tp": 3
		}
	}, {
		"comment": "",
		"default": "",
		"default_bit": null,
		"id": 15,
		"name": {
			"L": "b",
			"O": "b"
		},
		"offset": 12,
		"origin_default": "",
		"state": 5,
		"type": {
			"Charset": "binary",
			"Collate": "binary",
			"Decimal": 0,
			"Elems": null,
			"Flag": 4225,
			"Flen": -1,
			"Tp": 251
		}
	}],
	"comment": "",
	"id": 330,
	"index_info": [],
	"is_common_handle": false,
	"name": {
		"L": "test",
		"O": "test"
	},
	"partition": null,
	"pk_is_handle": true,
	"schema_version": 465,
	"state": 5,
	"update_timestamp":999 
}
```

<del>However, if we run `CREATE TABLE test ( id int(16) NOT NULL AUTO_INCREMENT, b longblob not null, PRIMARY KEY (id));` in TiDB, we can only get a JSON as below, which "default"/"origin_default" is `null` instead of `""`. And this won't make any error.</del>

### 2. What did you expect to see? (Required)

Table create successfully

### 3. What did you see instead (Required)

TiFlash crashes with `basic_string::_M_replace_aux`

### 4. What is your TiFlash version? (Required)
v5.2.1

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #4596

Problem Summary:

### What is changed and how it works?
make sure `Flen >= 0` before `append`.  

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
Two remaining problems:
1. do we need to support unsigned data structure?
2. for the TODO in `case TypeBit {}`, the current behavior is in line with expectations.

File: dbms/src/Databases/test/gtest_database.cpp [hunk 1]
```diff
@@ -688,7 +737,7 @@ try
     DatabaseLoading::loadTable(ctx, *db, meta_path, db_name, db_data_path, "TiFlash", "t_45.sql", false);
 
     // Get storage from database
-    const auto tbl_name = "t_45";
+    const auto * const tbl_name = "t_45";
```
Comment (ywqzzy) [inline dbms/src/Databases/test/gtest_database.cpp:740]: The second const is not neccessary.

Comment (Lloyd-Pottiger) [reply dbms/src/Databases/test/gtest_database.cpp:740]: ok, I will fix.

Comment (Lloyd-Pottiger) [inline dbms/src/Databases/test/gtest_database.cpp:740]:
```suggestion
    const auto * tbl_name = "t_45";
```

File: dbms/src/Storages/Transaction/TiDB.cpp [hunk 2]
```diff
@@ -155,10 +155,11 @@ Field ColumnInfo::defaultValueToField() const
         auto v = value.convert<String>();
         if (hasBinaryFlag())
         {
-            // For binary column, we have to pad trailing zeros according to the specified type length.
+            // For some binary column(like varchar(20)), we have to pad trailing zeros according to the specified type length.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/TiDB.cpp:158]: Add some comment to explain why we should check flen

File: dbms/src/Databases/test/gtest_database.cpp [hunk 3]
```diff
@@ -652,6 +652,55 @@ try
...
+        interpreter.execute();
+    }
+
+    auto db = ctx.getDatabase(db_name);
+
+    const String tbl_name = "t_111";
+    {
+        /// Create table
+        ParserCreateQuery parser;
+        const String stmt = fmt::format("CREATE TABLE `{}`.`{}` ", db_name, tbl_name) + R"stmt( (`id` Int32,`b` String) Engine = DeltaMerge((`id`),'{"cols":[{"comment":"","default":null,"default_bit":null,"id":1,"name":{"L":"id","O":"id"},"offset":0,"origin_default":null,"state":5,"type":{"Charset":"binary","Collate":"binary","Decimal":0,"Elems":null,"Flag":515,"Flen":16,"Tp":3}},{"comment":"","default":"","default_bit":null,"id":15,"name":{"L":"b","O":"b"},"offset":12,"origin_default":"","state":5,"type":{"Charset":"binary","Collate":"binary","Decimal":0,"Elems":null,"Flag":4225,"Flen":-1,"Tp":251}}],"comment":"","id":330,"index_info":[],"is_common_handle":false,"name":{"L":"test","O":"test"},"partition":null,"pk_is_handle":true,"schema_version":465,"state":5,"update_timestamp":99999}'))stmt";
```
Comment (JaySon-Huang) [inline dbms/src/Databases/test/gtest_database.cpp:677]: Better break down the R"stmt(..." into multiple lines for better reading https://en.cppreference.com/w/cpp/language/string_literal

Comment (Lloyd-Pottiger) [reply dbms/src/Databases/test/gtest_database.cpp]: fix

File: dbms/src/Storages/Transaction/TiDB.cpp (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/Transaction/TiDB.cpp:158]: fix

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**